src/main/java/ca/on/oicr/pineryreports/reports/TableReport.java

```diff
@@ -0,0 +1,195 @@
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+
+import ca.on.oicr.pinery.client.HttpResponseException;
+import ca.on.oicr.pinery.client.PineryClient;
+import ca.on.oicr.pineryreports.data.ColumnDefinition;
+import ca.on.oicr.pineryreports.data.ReportFormat;
+import ca.on.oicr.pineryreports.reports.pdf.AddHeaderFooterEvent;
+
+import com.google.common.collect.Sets;
+import com.itextpdf.io.font.FontConstants;
+import com.itextpdf.kernel.events.IEventHandler;
+import com.itextpdf.kernel.events.PdfDocumentEvent;
+import com.itextpdf.kernel.font.PdfFont;
+import com.itextpdf.kernel.font.PdfFontFactory;
+import com.itextpdf.kernel.geom.PageSize;
+import com.itextpdf.kernel.pdf.PdfDocument;
+import com.itextpdf.kernel.pdf.PdfWriter;
+import com.itextpdf.layout.Document;
+import com.itextpdf.layout.element.Cell;
+import com.itextpdf.layout.element.Paragraph;
+import com.itextpdf.layout.element.Table;
+import com.itextpdf.layout.property.TextAlignment;
+
+/**
+ * A type of report that has one table containing all of the data
+ */
+public abstract class TableReport implements Report {
+
+  private static final String EMPTY = "";
+  private static final String COMMA = ",";
+  private static final String NEWLINE = "\n";
+  private static final String QUOTE = "\"";
+  
+  public TableReport() {
+    
+  }
+
+  @Override
+  public Collection<ReportFormat> getValidFormats() {
+    return Sets.newHashSet(ReportFormat.CSV, ReportFormat.PDF);
+  }
+
+  @Override
+  public ReportFormat getDefaultFormat() {
+    return ReportFormat.CSV;
+  }
+
+  @Override
+  public final void generate(PineryClient pinery, ReportFormat format, File outFile) throws HttpResponseException, IOException {
+    collectData(pinery);
+    
+    switch (format) {
+    case CSV:
+      generateCsv(outFile);
+      break;
+    case PDF:
+      generatePdf(outFile);
+      break;
+    default:
+      // should be unreachable
+      throw new IllegalArgumentException("Invalid report format: " + format);
+    }
+  }
+  
+  private void generateCsv(File outFile) throws IOException {
+    try (FileWriter writer = new FileWriter(outFile)) {
+      String[] headings = getColumnHeadings(getColumns());
+      writer.write(toCsvLine(headings));
+      writer.write(NEWLINE);
+      
+      for (int i = 0, count = getRowCount(); i < count; i++) {
+        String[] row = getRow(i);
+        writer.write(toCsvLine(row));
+        writer.write(NEWLINE);
+      }
+    }
+  }
+  
+  private String toCsvLine(String[] rowData) {
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; i < rowData.length; i++) {
+      if (i > 0) {
+        sb.append(COMMA);
+      }
+      sb.append(QUOTE);
+      sb.append(emptyIfNull(rowData[i]));
+      sb.append(QUOTE);
+    }
+    return sb.toString();
+  }
+  
+  private void generatePdf(File outFile) throws FileNotFoundException, IOException {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'java.io.FileNotFoundException' which is a subclass of 'java.io.IOException'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)